iroh-relay/src/quic.rs

```diff
@@ -0,0 +1,301 @@
+
+            info!("QUIC server bound on {bind_addr:?}");
+
+            let cancel = CancellationToken::new();
+            let cancel_accept_loop = cancel.clone();
+
+            let task = tokio::task::spawn(async move {
+            let mut set = JoinSet::new();
+            debug!("waiting for connections...");
+            loop {
+                tokio::select! {
+                    biased;
+                    _ = cancel_accept_loop.cancelled() => {
+                        break;
+                    }
+                    Some(res) = set.join_next(), if !set.is_empty() => {
+                        if let Err(err) = res {
+                            if err.is_panic() {
+                                panic!("task panicked: {:#?}", err);
+                            }
+                            warn!("connection failed: {err}");
+                        }
+                    }
+                    res = endpoint.accept() => match res {
+                        Some(conn) => {
+                             debug!("accepting connection from {:?}", conn.remote_address())       ;
+                             set.spawn(async move {
+                                 let remote_addr = conn.remote_address();
+                                 let res = handle_connection(conn).await;
+                                 if let Err(ref err) = res {
+                                     warn!(remote_address = ?remote_addr, "error handling connection {err:?}")
+                                 }
+                                 res
+                             });
+                        }
+                        None => {
+                            debug!("endpoint closed");
+                            break;
+                        }
+                    }
+                }
+            }
+            endpoint
+                .close(QUIC_ADDR_DISC_CLOSE_CODE, QUIC_ADDR_DISC_CLOSE_REASON);
+            endpoint.wait_idle().await;
+            debug!("quic endpoint has been shutdown.");
+        }.instrument(info_span!("quic-endpoint")),);
+            Ok(Self {
+                bind_addr,
+                cancel,
+                handle: AbortOnDropHandle::new(task),
+            })
+        }
+
+        /// Closes the underlying QUIC endpoint and the tasks running the
+        /// QUIC connections.
+        pub fn shutdown(&self) {
+            self.cancel.cancel();
+        }
+    }
+
+    /// A handle for the Server side of QUIC address discovery.
+    ///
+    /// This does not allow access to the task but can communicate with it.
+    #[derive(Debug, Clone)]
+    pub struct ServerHandle {
+        cancel_token: CancellationToken,
+    }
+
+    impl ServerHandle {
+        /// Gracefully shut down the quic endpoint.
+        pub fn shutdown(&self) {
+            self.cancel_token.cancel()
+        }
+    }
+
+    async fn handle_connection(conn: quinn::Incoming) -> Result<()> {
+        let connection = conn.await?;
```

Comment:
This will trigger all the garbage that gets sent to this socket and log them as a warning by the caller.  (Quinn itself doesn't document this very well, we have at least some explanation hidden somewhere so we can say "told you so": https://docs.rs/iroh-net/latest/iroh_net/endpoint/struct.Incoming.html#method.accept).

So perhaps we can create something like:
```rust
#[thiserror::Error]
enum ConnError {
    Incoming(quinn::ConnectionError),
    Real(#[from] anyhow::Error),
}
```
so that the caller can distinguish maybe only debug-log the incoming errors, and then map the error from the incoming.

Reply:
ooo okay while implementing this, I saw that we end up debug-logging all other errors that happen in this method anyway. So I'm just debug-logging all errors in this method and not returning anything.

Reply:
to circle back to this, I added back `Return<()>` to the `handle_connection` method, but now we:

- log all errors in this method at the "debug" level (no more `warn` for `Incoming` errors)
- when errors are no longer handled at the call site, we just pass the `Result` along to the join set
- at the `set.join_next()` site, we only take action if the error is a `panic`, in which case we call `panic!`, otherwise the error has already been logged & we ignore it.